### PR TITLE
Add Deepseek 3.1 & Deepseek 3.1 base

### PR DESCRIPTION
**Pull request by cheachu**
## Description of the change

This PR introduces the **DeepSeek-V3.1** model architecture to Keras Hub, fully compliant with the Keras 3 backend-agnostic design (TensorFlow, JAX, PyTorch).
Key architectural components implemented in this PR:
* **Multi-head Latent Attention (MLA):** Implemented the MLA absorption trick to significantly reduce KV cache size during inference without materializing full per-head Key/Value tensors.
* **DeepSeekMoE (Mixture-of-Experts):** Implemented the auxiliary-loss-free top-K routing mechanism using Sigmoid affinity scores. Crucially, the expert routing and computation loop has been fully vectorized using `ops.einsum` and batched kernel tensors, ensuring it is **100% XLA-compatible** (`jit_compile=True`) and avoids the severe graph bloat associated with iterating over 256 experts.
* **YaRN RoPE Scaling:** Added Yet another RoPE extensioN (YaRN) for effective long-context scaling.
* **RMSNorm & SwiGLU FFN:** Implemented precise RMS normalization (using float32 casting for numerical stability) and dense SwiGLU layers for the initial non-MoE transformer layers.
* **Tokenizer & CausalLM:** Added the BytePairTokenizer (based on the official DeepSeek implementation) along with the `CausalLM` and `CausalLMPreprocessor` wrappers for text generation.

## Reference

* **Research Paper:** [DeepSeek-V3 Technical Report (arXiv:2412.19437)](https://arxiv.org/abs/2412.19437)
* **Reference Implementation:** [DeepSeek-V3 Official GitHub Repository](https://github.com/deepseek-ai/DeepSeek-V3)
* **Resolves Issue:** #[2391](https://github.com/keras-team/keras-hub/issues/2391)

## Checklist

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: keras_hub/src/models/deepseek_v31/__init__.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+)
+from keras_hub.src.models.deepseek_v31.deepseek_v31_presets import (
+    backbone_presets,
+)
+from keras_hub.src.models.deepseek_v31.deepseek_v31_presets import (
+    preprocessor_presets,
+)
+from keras_hub.src.models.deepseek_v31.deepseek_v31_presets import (
+    tokenizer_presets,
+)
```
[the remark this answers is not in the document]
Comment (cheachu) [reply keras_hub/src/models/deepseek_v31/__init__.py:20]: Had tried that ; But the pre-commit run --all-files command reverts it to the same version as it is now. Hence would be following that

File: keras_hub/src/models/deepseek_v31/deepseek_v31_attention.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+            self.rope_max_wavelength
+            ** (ops.arange(0, dim, 2, dtype="float32") / dim)
+        )
+
+        if self.rope_scaling_factor <= 1.0:
+            return ops.cast(freqs, dtype), self.yarn_mscale
+
+        # Wavelength = 2π / freq. High-freq → small wavelength, low-freq →
+        # large wavelength. YaRN applies more scaling to low-freq dimensions.
+        import math
```
[the remark this answers is not in the document]
Comment (cheachu) [reply keras_hub/src/models/deepseek_v31/deepseek_v31_attention.py]: Cool :) 
Will do that!

**Comment by cheachu** [comment]
Hi @sachinprasadhs 👋 
Can you review this? :)

**Comment by cheachu** [comment]
Hi @sachinprasadhs , @divyashreepathihalli 
Any update  :)

**Comment by sachinprasadhs** [comment]
@cheachu , Apologies for the delay. Will take a look into this this week. Thank you.

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    DeepSeekV31RMSNorm,
+)
+
+
+def _deepseek_v31_kernel_initializer(stddev=0.006):
+    return keras.initializers.RandomNormal(stddev=stddev)
+
+
+@keras_hub_export("keras_hub.models.DeepSeekV31Backbone")
+class DeepSeekV31Backbone(Backbone):
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py:24]: Is there any difference in architecture for DeepSeek 3.1 and DeepSeek 3? 
If not, then can we follow DeepSeekV3Backbone and follow this convenrion everywhere.
Only during checkpoint conversion, we can map the exact checkpoint names.

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    """DeepSeek V31 core transformer backbone.
+
+    Implements the full DeepSeek-V3 architecture as described in
+    arXiv:2412.19437. The model uses Multi-head Latent Attention (MLA) for
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py:28]: Make this hyperlink with title and https link.

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        num_routed_experts=256,
+        num_shared_experts=1,
+        num_experts_per_tok=8,
+        first_k_dense_replace=3,
+        rope_max_wavelength=10000,
+        rope_scaling_factor=1.0,
+        yarn_original_max_position_embeddings=4096,
+        layer_norm_epsilon=1e-6,
+        dropout=0.0,
+        **kwargs,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py:128]: Let's avoid too many defaults, only take defaults which is absolutely necessary and is common across multiple checkpoints variants.

Comment (sachinprasadhs) [reply keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py:128]: Everything will be passed to the model from the checkpoint configs.

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                    num_experts_per_tok=num_experts_per_tok,
+                    use_moe=(i >= first_k_dense_replace),
+                    rope_max_wavelength=rope_max_wavelength,
+                    rope_scaling_factor=rope_scaling_factor,
+                    yarn_original_max_position_embeddings=yarn_original_max_position_embeddings,  # noqa: E501
+                    layer_norm_epsilon=layer_norm_epsilon,
+                    dropout=dropout,
+                    kernel_initializer=_deepseek_v31_kernel_initializer(
+                        stddev=0.02
+                    ),
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py:167]: Can we just default `_deepseek_v31_kernel_initializer` with 0.002 then?

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        first_k_dense_replace=3,
+        rope_max_wavelength=10000,
+        rope_scaling_factor=1.0,
+        yarn_original_max_position_embeddings=4096,
+        layer_norm_epsilon=1e-6,
+        dropout=0.0,
+        **kwargs,
+    ):
+        dtype = kwargs.get("dtype")
+        # ===== Build sub-layers =====
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py:131]: Change `# ===== Build sub-layers =====` to `# ===== Layers =====`

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        super().__init__(
+            inputs={
+                "token_ids": token_id_input,
+                "padding_mask": padding_mask_input,
+            },
+            outputs=sequence_output,
+            **kwargs,
+        )
+
+        # ===== Store attributes (must be after super().__init__) =====
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py:201]: Change this to ` # === Config ===`

File: keras_hub/src/models/deepseek_v31/deepseek_v31_attention.py [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+        yarn_beta_fast=32,
+        yarn_beta_slow=1,
+        yarn_mscale=1.0,
+        yarn_mscale_all_dim=0.0,
+        yarn_original_max_position_embeddings=4096,
+        attention_dropout=0.0,
+        kernel_initializer="glorot_uniform",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_attention.py:117]: We can reduce lot of code by reusing our RotaryEmbedding with YaRN enabled here https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/layers/modeling/rotary_embedding.py
Please refactor the code after using existing `RotaryEmbedding`

File: keras_hub/src/models/deepseek_v31/deepseek_v31_attention.py [hunk 10]
```diff
@@ -0,0 +1,446 @@
+"""DeepSeek V31 Multi-head Latent Attention layer."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_attention.py:1]: Remove this docstring

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py [hunk 11]
```diff
@@ -0,0 +1,285 @@
+"""DeepSeek V31 backbone model."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone.py:1]: Remove this docstring

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone_test.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            "intermediate_dim": 128,
+            "q_lora_rank": 16,
+            "kv_lora_rank": 16,
+            "qk_nope_head_dim": 16,
+            "qk_rope_head_dim": 8,
+            "v_head_dim": 16,
+            "num_routed_experts": 4,
+            "num_shared_experts": 1,
+            "num_experts_per_tok": 2,
+            "first_k_dense_replace": 1,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone_test.py:27]: Let's test it with the smallest test configs possible.

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone_test.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        self.run_backbone_test(
+            cls=DeepSeekV31Backbone,
+            init_kwargs=self.init_kwargs,
+            input_data=self.input_data,
+            expected_output_shape=(2, 5, 64),
+        )
+
+    def test_num_parameters(self):
+        model = DeepSeekV31Backbone(**self.init_kwargs)
+        self.assertGreater(model.count_params(), 0)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone_test.py:44]: Let's test it with assertEqual with exact numer of parameters expected.

File: keras_hub/src/models/deepseek_v31/deepseek_v31_backbone_test.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        self.assertIsInstance(cache, list)
+        self.assertEqual(len(cache), self.init_kwargs["num_layers"])
+        for c_kv, k_rope in cache:
+            self.assertEqual(
+                c_kv.shape, (2, 5, self.init_kwargs["kv_lora_rank"])
+            )
+            self.assertEqual(
+                k_rope.shape, (2, 5, self.init_kwargs["qk_rope_head_dim"])
+            )
+
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_backbone_test.py:60]: You need to add `test_saved_model` also.

File: keras_hub/src/models/deepseek_v31/__init__.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
+"""DeepSeek V3.1 model exports."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/__init__.py:1]: Remove this

File: keras_hub/src/models/deepseek_v31/__init__.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    tokenizer_presets,
+)
+from keras_hub.src.models.deepseek_v31.deepseek_v31_tokenizer import (
+    DeepSeekV31Tokenizer,
+)
+from keras_hub.src.utils.preset_utils import register_presets
+
+register_presets(backbone_presets, DeepSeekV31Backbone)
+register_presets(tokenizer_presets, DeepSeekV31Tokenizer)
+register_presets(preprocessor_presets, DeepSeekV31CausalLMPreprocessor)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/__init__.py:28]: This is not required, remove it

File: keras_hub/src/models/deepseek_v31/__init__.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+)
+from keras_hub.src.models.deepseek_v31.deepseek_v31_presets import (
+    preprocessor_presets,
+)
+from keras_hub.src.models.deepseek_v31.deepseek_v31_presets import (
+    tokenizer_presets,
+)
+from keras_hub.src.models.deepseek_v31.deepseek_v31_tokenizer import (
+    DeepSeekV31Tokenizer,
+)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/__init__.py:23]: Remove these

File: keras_hub/src/models/deepseek_v31/__init__.py [hunk 18]
```diff
@@ -0,0 +1,28 @@
+"""DeepSeek V3.1 model exports."""
+
+from keras_hub.src.models.deepseek_v31.deepseek_v31_backbone import (
+    DeepSeekV31Backbone,
+)
+from keras_hub.src.models.deepseek_v31.deepseek_v31_causal_lm import (
+    DeepSeekV31CausalLM,  # noqa: F401
+)
+from keras_hub.src.models.deepseek_v31.deepseek_v31_causal_lm_preprocessor import (  # noqa: E501
+    DeepSeekV31CausalLMPreprocessor,
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/__init__.py:10]: Remove these lines.

File: keras_hub/src/models/deepseek_v31/deepseek_v31_causal_lm.py [hunk 19]
```diff
@@ -0,0 +1,182 @@
+"""DeepSeek V31 Causal Language Model."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_causal_lm.py:1]: Remove this

File: keras_hub/src/models/deepseek_v31/deepseek_v31_causal_lm.py [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+        preprocessor: Optional `DeepSeekV31CausalLMPreprocessor`. Used for
+            tokenizing inputs before passing to the model.
+
+    Example:
+
+    ```python
+    backbone = keras_hub.models.DeepSeekV31Backbone(...)
+    lm = keras_hub.models.DeepSeekV31CausalLM(backbone=backbone)
+    output = lm.generate("Once upon a time")
+    ```
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_causal_lm.py:44]: You can add from_preset example with the Keras Hub model name which will be uploaded after model gets merged.

File: keras_hub/src/models/deepseek_v31/deepseek_v31_causal_lm_preprocessor.py [hunk 21]
```diff
@@ -0,0 +1,62 @@
+"""DeepSeek V3.1 Causal LM Preprocessor."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_causal_lm_preprocessor.py:1]: Remove this

File: keras_hub/src/models/deepseek_v31/deepseek_v31_decoder_block.py [hunk 22]
```diff
@@ -0,0 +1,416 @@
+"""DeepSeek V31 transformer decoder block, RMSNorm, and dense FFN."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_decoder_block.py:1]: Remove this

File: keras_hub/src/models/deepseek_v31/deepseek_v31_presets.py [hunk 23]
```diff
@@ -0,0 +1,69 @@
+"""DeepSeek V3 model preset configurations."""
+
+# Metadata for loading pretrained model weights and configurations.
+backbone_presets = {
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_presets.py:4]: Comment this file and keep only empty dict for `backbone_presets`. This will be enabled only after we upload the checkpoints, which is after we merge the PR.

File: keras_hub/src/models/deepseek_v31/deepseek_v31_tokenizer.py [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+
+        super().__init__(vocabulary=vocabulary, merges=merges, **kwargs)
+
+        self._add_special_token("<｜begin▁of▁sentence｜>", "bos_token")
+        self._add_special_token("<｜end▁of▁sentence｜>", "eos_token")
+
+        self.start_token = "<｜begin▁of▁sentence｜>"
+        self.start_token_id = 151646
+        self.end_token_id = 151643
+        self.pad_token_id = 0
```
Comment (sachinprasadhs) [inline keras_hub/src/models/deepseek_v31/deepseek_v31_tokenizer.py:106]: Let's not hardcode these values here. These should be passed from tokenizer config file for respective mpdel checkpoints.
Refer other models.

**Review by sachinprasadhs — COMMENTED** [review]
Thank you for your patience. I have completed a high-level pass and added a few comments. For generic comments, please address them across the entire model implementation. Some parts of the code need to be aligned with the KerasHub standard, please refer to other models for guidance and make the necessary adjustments. The checkpoint conversion file under `tools/checkpoint_conversion` and the convert file under `keras_hub/src/utils/transformers` still need to be added. Once all these comments are addressed, I will do the review again. Thank you!

**Comment by sachinprasadhs** [comment]
This PR does not give credit by taking the commits from the original PR https://github.com/keras-team/keras-hub/pull/2171
And I'm curious to know on how you are palning to verify the numerics, since these models are very large 600+ Billion parameters.
